pkg/updater/resultstore/query_test.go

```diff
@@ -0,0 +1,663 @@
+			want:  "invocation_attributes.labels:\"foo\"",
+		},
+		{
+			query: "status:built special_type_target:some/special_type:target",
+			want:  "propertyEquals(special_type_target, \"some/special_type:target\")",
+		},
+		{
+			query: "label:\"foo\" special_type_target:some/special_type:target",
+			want:  "invocation_attributes.labels:\"foo\" propertyEquals(special_type_target, \"some/special_type:target\")",
+		},
+		{
+			query: "label:foo special_type_target:some/special_type:target target://my/test:target",
+			want:  "invocation.invocation_attributes.labels:\"foo\" (invocationPropertyEquals(special_type_target, \"some/special_type:target\") OR configurationPropertyEquals(special_type_target, \"some/special_type:target\")) id.target_id=\"//my/test:target\"",
+		},
+		{
+			query: "label:\"foo bar\" label:'baz baz'",
+			want:  "invocation_attributes.labels:\"foo bar\" invocation_attributes.labels:\"baz baz\"",
+		},
+		{
+			query: "exact_target://oh/no:bad label:foo",
+			want:  "id.target_id=\"//oh/no:bad\" invocation.invocation_attributes.labels:\"foo\"",
+		},
+		{
+			query: "target://oh/no:bad label:foo",
+			want:  "id.target_id=\"//oh/no:bad\" invocation.invocation_attributes.labels:\"foo\"",
+		},
+		{
+			query: "label:something SUBMIT_STATUS:POST_SUBMIT",
+			want:  "invocation_attributes.labels:\"something\" propertyEquals(SUBMIT_STATUS, \"POST_SUBMIT\")",
+		},
+		{
+			query: "label:foo label:bar bare-label",
+			want:  "invocation_attributes.labels:\"foo\" invocation_attributes.labels:\"bar\" invocation_attributes.labels:\"bare-label\"",
+		},
+		{
+			query: "label: bare-label",
+			want:  "invocation_attributes.labels:\"label:\" invocation_attributes.labels:\"bare-label\"",
+		},
+		{
+			query: "label: -bare-label",
+			want:  "invocation_attributes.labels:\"label:\" NOT invocation_attributes.labels:\"bare-label\"",
+		},
+		{
+			query: "label:foo empty:",
+			want:  "invocation_attributes.labels:\"foo\" invocation_attributes.labels:\"empty:\"",
+		},
+		{
+			query: "(label:foo label:bar)",
+			want:  "( invocation_attributes.labels:\"foo\" invocation_attributes.labels:\"bar\" )",
+		},
+		{
+			query: "label:\"foo (bar)\"",
+			want:  "invocation_attributes.labels:\"foo (bar)\"",
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.query, func(t *testing.T) {
+			got, err := ComplexQuery(tc.query)
+			if err != nil {
+				t.Errorf("complexQuery(%q) errored unexpectedly: %v", tc.query, err)
+			}
+			if got != tc.want {
+				t.Errorf("complexQuery(%q): got %q, want %q", tc.query, got, tc.want)
+			}
+		})
+	}
+}
+
+func TestQueryInvalid(t *testing.T) {
+	cases := []struct {
+		query string
+	}{
+		{"label:foo before:2020-01-01"},
+		{"after:2020-01-01"},
+	}
+	for _, tc := range cases {
+		t.Run(tc.query, func(t *testing.T) {
+			got, err := ComplexQuery(tc.query)
+			if err == nil {
+				t.Errorf("complexQuery(%q) expected error but got %q", tc.query, got)
+			}
+		})
+	}
+}
+
+type fakePage struct {
```

Comment:
maybe create a separate fake package or add this to an existing fakeclient package

Reply:
or add a TODO for that

Reply:
Hmm, we use both the convention of having it in the same package but separate file, and in a separate package, in this repo. Sticking it in the same package but separate file for now if that sounds good,, since it is specific to the ResultStore testing.